coda_v2_python_client/firebase_client_wrapper.py

```diff
@@ -367,3 +368,71 @@ def set_segment_messages_metrics(self, segment_id, messages_metrics):
+
+    def compute_segment_coding_progress(self, segment_id, messages=None):
+        """
+        Compute and return the progress metrics for a given dataset.
+
+        This method will initialise the counts in Firestore if they do not already exist.
+
+        :param segment_id: Id of a segment.
+        :type segment_id: str
+        :param messages: list of core_data_modules.data_models.message.Message, defaults to None
+        :param messages: If specified, it computes progress metrics based on the provided messages
+                         else it downloads messages from the requested segment. Defaults to None.
+        :type messages: core_data_modules.data_models.message.Message | None
+        :return: Messages metrics.
+        :rtype: core_data_modules.data_models.metrics.MessagesMetrics
+        """
+        if messages is None:
+            messages = self.get_segment_messages(segment_id)
+        messages_with_labels = 0
+        wrong_scheme_messages = 0
+        not_coded_messages = 0
+
+        code_schemes = {code_scheme.scheme_id: code_scheme for code_scheme in self.get_all_code_schemes(segment_id)}
+
+        for message in messages:
+            # Get the latest label from each scheme
+            latest_labels = {label.scheme_id: label for label in message.get_latest_labels()} # of scheme id -> label
+
+            # Test if the message has a label (that isn't SPECIAL-MANUALLY_UNCODED), and
+            # if any of the latest labels are either WS or NC
+            message_has_label = False
+            message_has_ws = False
+            message_has_nc = False
+            for label in latest_labels.values():
+                if label.code_id == "SPECIAL-MANUALLY_UNCODED":
+                    continue
+
+                if not label.checked:
+                    continue
+
+                message_has_label = True
+                scheme_for_label = code_schemes[label.scheme_id]
+                code_for_label = None
+
+                for code in scheme_for_label.codes:
+                    if label.code_id == code.code_id:
+                        code_for_label = code
+                assert code_for_label is not None
+
+                if code_for_label.code_type == "Control":
+                    if code_for_label.control_code == "WS":
+                        message_has_ws = True
+                    if code_for_label.control_code == "NC":
+                        message_has_nc = True
+
+            # Update counts appropriately
+            if message_has_label:
+                messages_with_labels += 1
+            if message_has_ws:
+                wrong_scheme_messages += 1
+            if message_has_nc:
+                not_coded_messages += 1
+
+        messages_metrics = MessagesMetrics(len(messages), messages_with_labels, wrong_scheme_messages, not_coded_messages)  # nopep8
+
+        # Write the metrics back if they weren't stored
```

Comment:
```suggestion
```
I think it's obvious what's happening here (but if you do want the comment, remove the 'if' section)